File: compiler/src/test/java/edu/cornell/cs/apl/viaduct/algebra/FreeDistributiveLatticeTest.java
package edu.cornell.cs.apl.viaduct.algebra;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class FreeDistributiveLatticeTest {
  private final FreeDistributiveLattice<String> elemA = FreeDistributiveLattice.create("a");
  private final FreeDistributiveLattice<String> elemB = FreeDistributiveLattice.create("b");
  private final FreeDistributiveLattice<String> elemC = FreeDistributiveLattice.create("c");
  private final FreeDistributiveLattice<String> top = FreeDistributiveLattice.top();
  private final FreeDistributiveLattice<String> bottom = FreeDistributiveLattice.bottom();

  @Test
  void testImply() {
    /* for all y, the greatest x s.t. bottom & x <= y is top */
    assertEquals(bottom.imply(elemA), top);

    /* for all y, the greatest x s.t. y & x <= y is top */
    assertEquals(elemA.imply(elemA), top);

    assertEquals(elemA.imply(elemB), elemB);

    assertEquals(elemA.imply(elemB.meet(elemA)), elemB);

    assertEquals(elemB.join(elemA).imply(elemA.join(elemB)), top);

    assertEquals(elemA.join(elemB).imply(elemA), elemA);

    assertEquals(elemA.meet(elemB).imply(elemA), top);

    assertEquals(elemA.imply(top), top);

    assertEquals(elemA.imply(bottom), bottom);

    assertEquals(elemA.imply(elemA.join(elemB)), top);

    assertEquals(elemA.imply(elemB.meet(elemC)), elemA.imply(elemB).meet(elemA.imply(elemC)));

    assertEquals(elemA.meet(elemA.imply(elemB)), elemA.meet(elemB));

    assertEquals(elemB.meet(elemA.imply(elemB)), elemB);
  }
}

File: compiler/src/test/kotlin/edu/cornell/cs/apl/viaduct/algebra/FreeDistributiveLatticeTest.kt
package edu.cornell.cs.apl.viaduct.algebra

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class FreeDistributiveLatticeTest {
    private val elemA = FreeDistributiveLattice("a")
    private val elemB = FreeDistributiveLattice("b")
    private val elemC = FreeDistributiveLattice("c")
    private val top = FreeDistributiveLattice.bounds<String>().top
    private val bottom = FreeDistributiveLattice.bounds<String>().bottom

    @Test
    fun testImply() {
        /* For all y, the greatest x s.t. bottom & x <= y is top. */
        Assertions.assertEquals(top, bottom.imply(elemA))

        /* For all y, the greatest x s.t. y & x <= y is top. */
        Assertions.assertEquals(top, elemA.imply(elemA))
        Assertions.assertEquals(elemB, elemA.imply(elemB))
        Assertions.assertEquals(elemB, elemA.imply(elemB.meet(elemA)))
        Assertions.assertEquals(top, elemB.join(elemA).imply(elemA.join(elemB)))
        Assertions.assertEquals(elemA, elemA.join(elemB).imply(elemA))
        Assertions.assertEquals(top, elemA.meet(elemB).imply(elemA))
        Assertions.assertEquals(top, elemA.imply(top))
        Assertions.assertEquals(bottom, elemA.imply(bottom))
        Assertions.assertEquals(top, elemA.imply(elemA.join(elemB)))
        Assertions.assertEquals(elemA.imply(elemB).meet(elemA.imply(elemC)), elemA.imply(elemB.meet(elemC)))
        Assertions.assertEquals(elemA.meet(elemB), elemA.meet(elemA.imply(elemB)))
        Assertions.assertEquals(elemB, elemB.meet(elemA.imply(elemB)))
    }
}
